Sources/PackageGraph/PinsStore.swift

```diff
@@ -23,56 +24,43 @@ public final class PinsStore {
         /// The pinned state.
         public let state: CheckoutState
 
-        public init(
-            packageRef: PackageReference,
-            state: CheckoutState
-        ) {
+        public init(packageRef: PackageReference, state: CheckoutState) {
             self.packageRef = packageRef
             self.state = state
         }
     }
 
-    /// The schema version of the resolved file.
-    ///
-    /// * 1: Initial version.
-    static let schemaVersion: Int = 1
-
-    /// The path to the pins file.
-    private let pinsFile: AbsolutePath
-
-    /// The filesystem to manage the pin file on.
-    private var fileSystem: FileSystem
-
     private let mirrors: DependencyMirrors
 
-    /// The pins map.
-    public fileprivate(set) var pinsMap: PinsMap
+    /// storage
+    private let storage: PinsStorage
+    private let _pins: ThreadSafeKeyValueStore<PackageIdentity, PinsStore.Pin>
 
     /// The current pins.
-    public var pins: AnySequence<Pin> {
-        return AnySequence<Pin>(pinsMap.values)
+
+    public var pinsMap: PinsMap {
+        self._pins.get()
     }
 
-    fileprivate let persistence: SimplePersistence
+    public var pins: AnySequence<Pin> {
+        return AnySequence<Pin>(self.pinsMap.values)
+    }
 
     /// Create a new pins store.
     ///
     /// - Parameters:
     ///   - pinsFile: Path to the pins file.
     ///   - fileSystem: The filesystem to manage the pin file on.
     public init(pinsFile: AbsolutePath, fileSystem: FileSystem, mirrors: DependencyMirrors) throws {
-        self.pinsFile = pinsFile
-        self.fileSystem = fileSystem
-        self.persistence = SimplePersistence(
-            fileSystem: fileSystem,
-            schemaVersion: PinsStore.schemaVersion,
-            statePath: pinsFile,
-            prettyPrint: true)
-        self.pinsMap = [:]
+        self.storage = .init(path: pinsFile, fileSystem: fileSystem)
         self.mirrors = mirrors
+
         do {
-            _ = try self.persistence.restoreState(self)
-        } catch SimplePersistence.Error.restoreFailure(_, let error) {
+            self._pins = .init(try self.storage.load(mirrors: mirrors))
+        } catch {
+            self._pins = .init()
+            // FIXME: delete the file?
+            // FIXME: warning instead of error?
```

Comment:
It makes sense that this would be an error, and then the client could choose the policy (error, or catch the error and delete the file and try again, etc).

Reply:
right, in the case of the package.resolved file this may best - we let the user make the decision

@neonichu wdyt?